src/olympia/addons/serializers.py

```diff
@@ -211,6 +212,48 @@ class Meta:
                   'release_notes', 'reviewed', 'url', 'version')
 
 
+class CurrentVersionSerializer(SimpleVersionSerializer):
+    def to_representation(self, obj):
+        # If the add-on is a langpack, and `appversion` is passed, try to
+        # determine the latest public compatible version and replace the obj
+        # with the result. Because of the perf impact, only done for langpacks
+        # in the detail API.
+        request = self.context.get('request')
+        view = self.context.get('view')
+        addon = obj.addon
+        if (request and request.GET.get('appversion') and
+                getattr(view, 'action', None) == 'retrieve' and
+                addon.type == amo.ADDON_LPAPP):
+            obj = self.get_current_compatible_version(addon)
+        return super(CurrentVersionSerializer, self).to_representation(obj)
+
+    def get_current_compatible_version(self, addon):
+        """
+        Return latest public version compatible with the app & appversion
+        passed through the request, or fall back to addon.current_version if
+        none is found.
+
+        Only use on langpacks if the appversion parameter is present.
+        """
+        request = self.context.get('request')
+        try:
+            application = AddonAppQueryParam(request).get_value()
+        except ValueError:
+            raise exceptions.ParseError('Invalid or missing app parameter.')
+        try:
+            # AddonAppVersionQueryParam.get_values() returns (app_id, min, max)
+            # but we want {'min': min, 'max': max}.
+            appversions = dict(
+                zip(('min', 'max'),
+                    AddonAppVersionQueryParam(request).get_values()[1:]))
+        except ValueError:
+            raise exceptions.ParseError('Invalid appversion parameter.')
```

Comment:
Actually, why isn't `AddonAppVersionQueryParam` used?  It already has validation for app and appversion and raises an appropriate error message.

Reply:
Probably right, I'll check & remove the `AddonAppQueryParam` usage.